lib/Core/TargetForest.cpp

```diff
@@ -94,11 +166,37 @@ void TargetForest::Layer::block(ref<Target> target) {
   }
 }
 
+void TargetForest::Layer::removeTarget(ref<Target> target) {
+  auto it = targets2Vector.find(target);
+
+  if (it == targets2Vector.end()) {
+    return;
+  }
+
+  auto targetsVectors = std::move(it->second);
+
+  targets2Vector.erase(it);
+
+  for (auto& targetsVec : targetsVectors) {
+    bool shouldDelete = true;
+
+    for (auto& target_i : targetsVec->getTargets()) {
+      if (targets2Vector.count(target_i) != 0) {
+        shouldDelete = false;
+      }
+    }
+
+    if (shouldDelete) {
+      forest.erase(targetsVec);
+    }
+  }
+}
+
 bool TargetForest::Layer::deepFind(ref<Target> target) const {
   if (empty())
     return false;
-  auto res = forest.find(target);
-  if (res != forest.end()) {
+  auto res = targets2Vector.find(target);
```

Comment:
Please do not replace words with numbers in the names of functions, classes, fields, etc.